File: web/core/auth/phone-enforcement.ts
import type { AuthProfile } from "@/core/api/types";

export type PhoneEnforcementMode = "soft" | "warning" | "strict";

const PHONE_EXEMPT_ROLES = new Set(["SUPER_ADMIN", "CHURCH_ADMIN"]);

const ENFORCEMENT_STATUSES = new Set(["ACTIVE", "PENDING"]);

const STRICT_ALLOWED_ROUTES = new Set([
  "/dashboard",
  "/dashboard/settings/profile",
  "/dashboard/finance/my-contributions",
  "/dashboard/notifications",
]);

function normalizeRoute(route: string): string {
  const withoutQuery = route.split("?")[0] ?? route;
  if (withoutQuery.length > 1 && withoutQuery.endsWith("/")) {
    return withoutQuery.slice(0, -1);
  }
  return withoutQuery || "/dashboard";
}

export function isPhoneEnforcementExempt(profile: AuthProfile | null): boolean {
  if (!profile) {
    return true;
  }
  return profile.roles.some((role) => PHONE_EXEMPT_ROLES.has(role));
}

export function memberMissingPhone(profile: AuthProfile | null): boolean {
  if (!profile?.member || isPhoneEnforcementExempt(profile)) {
    return false;
  }

  const status = profile.member.status;
  if (!status || !ENFORCEMENT_STATUSES.has(status)) {
    return false;
  }

  return !profile.member.phone;
}

export function getPhoneEnforcementMode(
  profile: AuthProfile | null,
): PhoneEnforcementMode {
  if (!profile?.phoneEnforcement?.enabled) {
    return "soft";
  }
  return profile.phoneEnforcement.mode ?? "soft";
}

export function isStrictPhoneBlocked(profile: AuthProfile | null): boolean {
  if (!profile?.phoneEnforcement?.enabled) {
    return false;
  }
  if (profile.phoneEnforcement.mode !== "strict") {
    return false;
  }
  return memberMissingPhone(profile);
}

export function requiresPhone(profile: AuthProfile | null): boolean {
  return memberMissingPhone(profile);
}

export function canAccessRoute(
  profile: AuthProfile | null,
  route: string,
): boolean {
  if (!isStrictPhoneBlocked(profile)) {
    return true;
  }

  const normalized = normalizeRoute(route);
  if (normalized === "/dashboard") {
    return true;
  }

  for (const allowed of STRICT_ALLOWED_ROUTES) {
    if (allowed === "/dashboard") {
      continue;
    }
    if (normalized === allowed || normalized.startsWith(`${allowed}/`)) {
      return true;
    }
  }

  return false;
}

export function phoneEnforcementRedirectRoute(): string {
  return "/dashboard/settings/profile";
}

export function shouldHideOperationalNavigation(
  profile: AuthProfile | null,
): boolean {
  return isStrictPhoneBlocked(profile);
}

export function isWarningPhoneMode(profile: AuthProfile | null): boolean {
  return (
    getPhoneEnforcementMode(profile) === "warning" && memberMissingPhone(profile)
  );
}

export function isSoftPhoneMode(profile: AuthProfile | null): boolean {
  return (
    getPhoneEnforcementMode(profile) === "soft" && memberMissingPhone(profile)
  );
}

File: mobile/lib/core/auth/phone_enforcement.dart
import '../routing/app_router.dart';

const _phoneExemptRoles = {'SUPER_ADMIN', 'CHURCH_ADMIN'};

const _enforcementStatuses = {'ACTIVE', 'PENDING'};

const _strictAllowedRoutes = {
  AppRouter.memberDashboard,
  AppRouter.leaderDashboard,
  AppRouter.operational,
  AppRouter.settings,
  AppRouter.language,
  AppRouter.notifications,
  AppRouter.myContributions,
};

class PhoneEnforcementState {
  const PhoneEnforcementState({
    required this.enabled,
    required this.mode,
    required this.blocked,
  });

  final bool enabled;
  final String mode;
  final bool blocked;
}

PhoneEnforcementState? parsePhoneEnforcement(Map<String, dynamic>? profile) {
  final raw = profile?['phoneEnforcement'];
  if (raw is! Map) {
    return null;
  }

  return PhoneEnforcementState(
    enabled: raw['enabled'] == true,
    mode: raw['mode']?.toString() ?? 'soft',
    blocked: raw['blocked'] == true,
  );
}

List<String> roleNamesFromProfile(Map<String, dynamic>? profile) {
  final fromApi = profile?['roles'] as List?;
  if (fromApi != null) {
    return fromApi.map((entry) => entry.toString()).toList();
  }

  final roles = profile?['userRoles'] as List?;
  if (roles == null) {
    return const [];
  }

  return roles
      .map((entry) => (entry as Map)['role']?['name']?.toString())
      .whereType<String>()
      .toList();
}

bool isPhoneEnforcementExempt(Map<String, dynamic>? profile) {
  return roleNamesFromProfile(profile)
      .any((role) => _phoneExemptRoles.contains(role));
}

bool memberMissingPhone(Map<String, dynamic>? profile) {
  if (profile == null || isPhoneEnforcementExempt(profile)) {
    return false;
  }

  final member = profile['member'] as Map?;
  if (member == null) {
    return false;
  }

  final status = member['status']?.toString();
  if (status == null || !_enforcementStatuses.contains(status)) {
    return false;
  }

  final phone = member['phone']?.toString();
  return phone == null || phone.isEmpty;
}

bool requiresPhone(Map<String, dynamic>? profile) {
  return memberMissingPhone(profile);
}

bool isStrictPhoneBlocked(Map<String, dynamic>? profile) {
  final enforcement = parsePhoneEnforcement(profile);
  if (enforcement == null || !enforcement.enabled) {
    return false;
  }
  if (enforcement.mode != 'strict') {
    return false;
  }
  return requiresPhone(profile);
}

bool canAccessRouteWithPhoneEnforcement(
  String routeName,
  Map<String, dynamic>? profile,
) {
  if (!isStrictPhoneBlocked(profile)) {
    return true;
  }
  return _strictAllowedRoutes.contains(routeName);
}

String phoneEnforcementRedirectRoute() {
  return AppRouter.settings;
}
